contributors/design-proposals/core-metrics-pipeline.md

```diff
@@ -0,0 +1,295 @@
+  Memory *MemoryResources `json:"memory,omitempty"`  
+  // Stats pertaining to node filesystem resources.  
+  // +optional  
+  Filesystems []FilesystemResources `json:"filesystems, omitempty" patchStrategy:"merge" patchMergeKey:"device"`  
+}  
+
+// CpuResources containes data about cpu resource usage  
+type CpuResources struct {  
+  // The number of cores in this machine.  
+  NumCores int `json:"numcores"`  
+  // The current Usage of CPU resources  
+  Usage *CpuUsage `json:"cpuusage,omitempty"`  
+}  
+
+// MemoryResources contains data about memory resource usage.  
+type MemoryResources struct {  
+  // The time at which these stats were updated.  
+  Timestamp metav1.Time `json:"time"`  
+  // The memory capacity, in bytes  
+  CapacityBytes *uint64 `json:"capacitybytes,omitempty"`  
+  // The available memory, in bytes  
+  // This is the number of bytes which are not included in the working set memory
+  // Working set memory includes recently accessed memory, dirty memory, and kernel memory.
+  AvailableBytes *uint64 `json:"availablebytes,omitempty"`  
+}  
+
+// FilesystemResources contains data about filesystem disk resources.  
+type FilesystemResources struct {  
+  // The time at which these stats were updated.  
+  Timestamp metav1.Time `json:"time"`  
+  // The device that this filesystem is on  
+  Device string `json:"device"`  
+  // AvailableBytes represents the storage space available (bytes) for the filesystem.  
+  // +optional  
+  AvailableBytes *uint64 `json:"availableBytes,omitempty"`  
+  // CapacityBytes represents the total capacity (bytes) of the filesystems underlying storage.  
+  // +optional  
+  CapacityBytes *uint64 `json:"capacityBytes,omitempty"`  
+  // InodesFree represents the free inodes in the filesystem.  
+  // +optional  
+  InodesFree *uint64 `json:"inodesFree,omitempty"`  
+  // Inodes represents the total inodes in the filesystem.  
+  // +optional  
+  Inodes *uint64 `json:"inodes,omitempty"`  
+}  
+
+// PodUsage holds pod-level unprocessed sample stats.  
+type PodUsage struct {  
+  // UID of the pod  
+  PodUID string `json:"uid"`  
+  // Stats pertaining to pod total usage of cpu  
+  // This may include additional overhead not included in container usage statistics.  
+  // +optional  
+  CPU *CpuUsage `json:"cpu,omitempty"`  
+  // Stats pertaining to pod total usage of system memory  
+  // This may include additional overhead not included in container usage statistics.  
+  // +optional  
+  Memory *MemoryUsage `json:"memory,omitempty"`  
+  // Stats of containers in the pod.  
+  Containers []ContainerUsage `json:"containers" patchStrategy:"merge" patchMergeKey:"uid"`  
+  // Stats pertaining to volume usage of filesystem resources.  
+  // +optional  
+  Volumes []VolumeUsage `json:"volume,omitempty" patchStrategy:"merge" patchMergeKey:"name"`  
+}  
+
+// ContainerUsage holds container-level usage stats.  
+type ContainerUsage struct {  
+  // UID of the container  
```

Comment:
nit: I don't think we ever refer to the container by UID?  The CRI calls it the "ContainerID".

Can the master components reference containers by UID? Or do they need container name to track the container across time (restarts)? @davidopp @timothysc